src/platform/backends/shared/base_virtual_machine.cpp

```diff
@@ -847,3 +822,44 @@ void mp::BaseVirtualMachine::drop_ssh_session()
         ssh_session.reset();
     }
 }
+
+auto mp::BaseVirtualMachine::try_to_ssh() -> utils::TimeoutAction
+{
+    detect_aborted_start();
+    refresh_start();
+    try
+    {
+        ssh_and_cross_to_running();
+        return utils::TimeoutAction::done;
+    }
+    catch (const InternalTimeoutException& e)
+    {
+        return log_and_retry(e, this);
+    }
+    catch (const SSHException& e)
+    {
+        return log_and_retry(e, this);
+    }
+    catch (const IPUnavailableException& e)
+    {
+        return log_and_retry(e, this);
+    }
+}
+
+void mp::BaseVirtualMachine::ssh_and_cross_to_running()
+{
+    static constexpr auto wait_step = 1s;
+    ssh_session.emplace(ssh_hostname(wait_step), ssh_port(), ssh_username(), key_provider);
+
+    std::lock_guard lock{state_mutex};
```

Comment:
I wonder if we should extract `lock-set state-update` to its own function for convenience.